src/test/groovy/com/commercehub/gradle/plugin/avro/OptionsFunctionalSpec.groovy

```diff
@@ -208,17 +205,24 @@ class OptionsFunctionalSpec extends FunctionalSpec {
         result.task(":generateAvroJava").outcome == SUCCESS
         def content = projectFile("build/generated-main-avro-java/example/avro/User.java").text
 
-        and: "the specified createOptionalGetters is used"
-        content.contains("public Optional<java.lang.String> getFavoriteColor()") == expectedPresent
+        and: "the specified optionalGettersForNullableFieldsOnly is used"
+        content.contains("public Optional<java.lang.String> getFavoriteColor()") == expectedNullableOptionalGetter
+        content.contains("public Optional<java.lang.String> getName()") == expectedRequiredOptionalGetter
```

Comment:
```suggestion
        expectedNullableOptionalGetter ? content.contains("public Optional<java.lang.String> getFavoriteColor()") : content.contains("public java.lang.String getFavoriteColor()")
        expectedRequiredOptionalGetter ? content.contains("public Optional<java.lang.String> getName()") : content.contains("public java.lang.String getName()")
```

Reply:
I've implemented this change but want to clarify my original comment as I'm not sure this test criteria has the intended behavior.

The inner builder that avro generates for the POJOs will always contain an unmodified `public <type> getX()` regardless of the options relating to top level getters. As a result the following condition in the above test case will pass for every configuration of both `gettersReturnOptional` and `optionalGettersForNullableFieldsOnly`

```
content.contains("public java.lang.String getFavoriteColor()")
content.contains("public java.lang.String getName()")
```

The existing tests were attempting to ensure that Optional getters were not created rather than testing for the existence of the standard getters as the standard getters always exist on a naive contains (due to the inner builder class).

So the condition of `content.contains("public java.lang.String getX()")` isn't going to provide any greater testing confidence than omitting that line entirely (as it will always be true).

Reply:
Ooh. Good point. I’ll come up with some way to correct for that.

Reply:
I think the way to go is along the lines of the regex suggestion you originally made.  Clear method and variable naming will make it clear enough to a casual reader.  It may break occasionally when they change their templates, but in most cases it should break in a way that causes false test failures, rather than false test passes (which are worse).  I think the regex is less likely to break than something based on line positions or match counts.  I also think that this sort of quick approach is better than pulling in a dependency on some form of source or bytecode parsing library, as the complexity isn't warranted for this use case.

What I'm thinking is along the lines of having each method that currently has assertions against `content` include a line immediately after loading `content`: `def mainClassContent = getMainClassContent(content)`.  The method would look like this:

```groovy
    /**
     * Returns just the portion of a file that relates to the main class.
     * This is used in order to allow assertions on the getters/setters/fields of the generated class itself, as opposed to a Builder.
     * 
     * @param content the file content for which to get the main content
     * @return the content of the class, from the start of the class body to the first inner class definition
     */
    @SuppressWarnings("LineLength")
    private static String getMainClassContent(String content) {
        def className = "User"
        def matcher = content =~ /(?s)public class ${className} extends org\.apache\.avro\.specific\.\SpecificRecordBase implements org\.apache\.avro\.specific\.SpecificRecord \{(?<mainClassContent>.*)public static class Builder/
        assert matcher.find()
        return matcher.group("mainClassContent")
    }
```

And then the assertions would use `mainClassContent` rather than `content`, sort of like this:

```groovy
        mainClassContent.contains("public Optional<java.lang.String> getFavoriteColor()") == expectedNullableOptionalGetter
        mainClassContent.contains("public java.lang.String getFavoriteColor()") != expectedNullableOptionalGetter
        mainClassContent.contains("public Optional<java.lang.String> getName()") == expectedRequiredOptionalGetter
        mainClassContent.contains("public java.lang.String getName()") != expectedRequiredOptionalGetter
```

Reply:
I can make this change as part of this PR, I may not get to it for a few days but since we're waiting on the Avro release anyways seems like a good place to do it.

Reply:
Sounds good. Please do. No hurry. As you say, we’re going to be blocked waiting for the release any way. Thanks for your work on this. You’ve been a big help.

Reply:
I didn't realize you had written an accurate regex and properly compiling method. This was essentially a drop in, though I changed any other tests that concerned themselves specifically with conditions of the main body to use `mainClassContent` as well.